datarobot_batch_scoring/batch_scoring.py

```diff
@@ -41,53 +45,126 @@ class ShelveError(Exception):
     REGRESSION = 'Regression'
     BINARY = 'Binary'
 
 
+def fast_to_csv_chunk(data, header):
+    """Fast routine to format data for prediction api.
+
+    Returns data in unicode.
+    """
+    header = ','.join(header)
+    return ''.join(chain((header, '\n'), data))
+
+
+def slow_to_csv_chunk(data, header):
+    """Slow routine to format data for prediction api.
+
+    Returns data in unicode.
+    """
+    if six.PY3:
+        buf = io.StringIO()
+    else:
+        buf = io.BytesIO()
+
+    writer = csv.writer(buf)
+    writer.writerow(header)
+    writer.writerows(data)
+    return buf.getvalue()
+
+
+class FastReader(object):
+    """A reader that only reads the file in text mode but not parses it. """
+
+    def __init__(self, fd, sep=",", dialect=None, encoding='utf-8'):
+        self.fd = fd
+        reader = csv.reader(self.fd, dialect, delimiter=sep)
+        self.header = next(reader)
+        self.fieldnames = [c.strip() for c in self.header]
+
+    def __iter__(self):
+        self.fd.seek(0)
+        it = iter(self.fd)
+        next(it)  # skip header
+        return it
+
+
+class SlowReader(object):
+    """The slow reader does actual CSV parsing.
+    It supports multiline csv and can be a factor of 50 slower. """
+
+    def __init__(self, fd, sep=",", dialect=None, encoding='utf-8'):
+        self.fd = fd
+        self.sep = sep
+        self.dialect = dialect
+        self.reader = csv.reader(self.fd, dialect, delimiter=sep)
+        self.header = next(self.reader)
+        self.fieldnames = [c.strip() for c in self.header]
+
+    def __iter__(self):
+        self.fd.seek(0)
+        self.reader = csv.reader(self.fd, self.dialect, delimiter=self.sep)
+        for i, row in enumerate(self.reader):
+            if i == 0:
+                # skip header
+                continue
+            yield row
+
+
 class BatchGenerator(object):
     """Class to chunk a large csv files into a stream
     of batches of size ``--n_samples``.
 
     Yields
     ------
     batch : Batch
-        The next batch
+        The next batch. A batch holds the data to be send already
+        in the form that can be passed to the HTTP request.
     """
 
-    def __init__(self, dataset, n_samples, n_retry, delimiter, ui):
+    def __init__(self, dataset, n_samples, n_retry, delimiter, ui, fast_mode):
         self.dataset = dataset
         self.chunksize = n_samples
         self.rty_cnt = n_retry
         self.sep = delimiter
         self._ui = ui
+        self.fast_mode = fast_mode
+        self.encoding = 'utf-8'
 
-    def open(self):
+    def open(self, encode=True):
+        if six.PY3:
+            mode = 'rt'
+        else:
+            mode = 'rb'
         if self.dataset.endswith('.gz'):
-            return gzip.open(self.dataset)
+            fd = gzip.open(self.dataset, mode)
         else:
             if six.PY2:
-                return open(self.dataset, 'rb')
+                fd = open(self.dataset, mode)
```

Comment:
I don't see difference between six.PY2 and PY3, they looks the same:
`fd = open(self.dataset, mode)`
and
`fd = open(self.dataset, mode)`